README.md
PCE - PC Emulator
==============================================================================

PCE is a collection of microcomputer emulators.'
This is a fork from Hampa Hug <hampa@hampa.ch> https://github.com/retrohun/pce

My purpose for forking this was to make modifications to support creating a working 1/3 scale Macintosh SE using a Raspberry Pi Zero in a hand crafted case.

Install
==============================================================================
The usual...

./configure

make

sudo make install

Based on this blog post by ToughDev, http://www.toughdev.com/content/2016/11/pcemacplus-the-ultimate-68k-classic-macintosh-emulator/, I used the following configuration options:

--with-x 

--with-sdl 

--enable-tun 

--enable-char-ppp 

--enable-char-tcp 

--enable-sound-oss 

--enable-char-slip 

--enable-char-pty 

--enable-char-posix 

--enable-char-termios 

CFLAGS=-I/usr/local/include/SDL2

Run
==============================================================================
pce-macplus -v -c mac-se.cfg -l pce.log -r

Where  

-c mac-se.cfg loads the config file mac-se.cfg

-r runs the emulator

-l pce.log writes to the log file pce.log

-v logs verbose comments


ChangeLog
commit ba51ca7dde64d2b487673405ea90a716938be5c8
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-03-15 12:58:32 +0100

    Update for version 0.2.2

commit be14b57735f84291c57f55366d65cb854bc61dae
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-03-15 12:56:42 +0100

    configure: Regenerate configure with autoconf 2.69

commit cb9ef43ff7ba4448d1f92cf5c5d922960816e939
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-03-15 12:55:50 +0100

    Update autoconf helper scripts

commit 676267ca02a570b7b243710a266ac15d8d69d7ea
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-18 17:38:37 +0100

    hgc: Fix scrolling in HGC graphic mode

commit d0220e9c8d6feac14e482de2f035abbd6047336a
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-05 19:39:55 +0100

    xt-hdc: Clear the data direction bit after reading the command result

commit 404a13feec992801be8975d3b7b5ce7328dcdeeb
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-03 11:47:58 +0100

    ega: Fix line compare

commit abb47c0d89787096c208f72c95dcc086c2f40971
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-03 11:47:25 +0100

    Rename src/utils/pceimg/ to src/utils/pce-img/

commit c9a52e7db20bac2ad90e121e32687d096beb651d
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-03 05:52:08 +0100

    qed: Report unknown features

commit 28748dd93e00724d83fa2fd1003436a9840fdf3b
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-02-03 05:51:26 +0100

    pfdc: Fix loading alternate sectors from pfdc version 4 files

commit 3a9b34117f5e0cdddac7163fd5b354486f711cec
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-30 22:08:41 +0100

    video: Add message "term.escape" to set the escape key at runtime

commit d4b47185ddc32351dabcc411294e0848e87eb1dd
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-30 22:08:09 +0100

    8272-fdc: Don't abort the read track command after a ID CRC error

commit e679ff0626f05925bf5f8d33bd8ce86739a69012
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-11 16:34:32 +0100

    lib: Move the 'h' command to the monitor

commit 831e622c7acc465df787183499a2cdd1d935c9e6
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-11 14:24:28 +0100

    monitor: Add the 'save' command instead of the 'w' command
    
    The save command saves emulated memory to ihex, srec or
    binary files.

commit ae9ca37367f88e98a1ef9d2094f446d1c37e5700
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-11 14:24:00 +0100

    monitor: Add the 'load' command
    
    The load command loads ihex, srec or binary files into the
    emulated memory.

commit 459dd0a1f0adeb6ab64c90a63b8d1cb7e58b6825
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-11 14:21:23 +0100

    simarm: Refactor include files

commit 01c6c8aa08a3974c039ea94ba6efe559dc174014
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-10 11:42:10 +0100

    lib: Automatically detect the file type for files loaded into memory

commit 1b0fc7a8bb23e7b1808e68011ea7ee28392c4dce
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-10 11:41:42 +0100

    lib: Support saving segmented IHEX files

commit 1ebaee4e9e1cff8d10076d9cd5671f341d8421d7
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-10 11:40:54 +0100

    lib: Fix an endless loop when saving SREC files

commit 4b81710be4d1c1752d1c3319aaabdbd6d87e4ebf
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-10 11:40:28 +0100

    Fix documentation for the mouse character driver

commit 9cf39d408d52e64c9acd62e096501d1cad21e2d7
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-06 02:46:21 +0100

    video: Fix MDA memory addressing and cursor display
    
    Although the MDA's memory is only 4096 bytes, memory accesses beyond
    that need to wrap around.
    
    This patch also fixes a long standing bug that prevented the cursor
    from being displayed correctly if the displayed part of memory wrapped
    around from 4095 to 0.

commit 939e1a0303946e35331c0d6104647967e5b1da4b
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-06 02:45:50 +0100

    char: Support Mouse Systems serial mice

commit f51bdd2c6a44d13be6f4ca1270dc0c41e7c8ea4f
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-03 13:17:46 +0100

    brkpt: Unify address breakpoints

commit d70a8f765c7842cb3d02632d516d2a6df5313d15
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-03 13:17:13 +0100

    e8086: Disable interrupts after SS is loaded

commit bea9c8f8ac3ba2980d612319c94b78ce48b72903
Author: Hampa Hug <hampa@hampa.ch>
Date:   2013-01-03 13:16:44 +0100

    e8086: Add a prefetch delay after jump instructions

commit 2affd58b9b66708099f6ee78183efa084c6455f6
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-22 12:10:05 +0100

    e8259: Clear the interrupt mask register during initialization

commit 43167998ee72d7a989862c720aecdc6b157118a5
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-22 12:09:38 +0100

    sysdep: Trun off XON/XOFF flow control for raw streams

commit 8b19403ff7dca4901783d25828b0927b6f12b0d5
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-08 20:35:55 +0100

    sysdep: Don't generate signals when a file descriptor is non-interactive

commit b961fbd4e21cc47faf2eb42d557119167a743642
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-08 20:34:49 +0100

    pfdc: Support creating new CP/M images
    
    Creating new CP/M images is really the same as creating DOS
    images, except that the new images are initialized with E5
    instead of F6.

commit ddccbc50f18e98ee243b4d5dc6cf6f81e4762fce
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-08 20:34:09 +0100

    pfdc: Fix encoding of 8 inch SSSD disks

commit 963c763106096777b2c35f806c9b8fd84b13154f
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-08 20:33:05 +0100

    pfdc: Probe raw images and use them if they match a known image size

commit 71c5ee9776bdabb17ce24e84fbc30a74ee2414bb
Author: Hampa Hug <hampa@hampa.ch>
Date:   2012-12-08 20:32:37 +0100

    pfdc: Allow round sizes when guessing the disk geometry from the disk size

commit d91084b03396d35c7ed6621a6b257ec9f847ef14

src/config.h
/* src/config.h.  Generated from config.h.in by configure.  */
/*****************************************************************************
 * pce                                                                       *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/config.h.in                                              *
 * Created:     2002-12-16 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2002-2011 Hampa Hug <hampa@hampa.ch>                     *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#ifndef PCE_CONFIG_H
#define PCE_CONFIG_H 1


#ifndef _GNU_SOURCE
#define _GNU_SOURCE 1
#endif


#ifndef _NETBSD_SOURCE
#define _NETBSD_SOURCE 1
#endif


#define PCE_LARGE_FILE 1

#ifdef PCE_LARGE_FILE
#ifndef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif
#endif


#define HAVE_FCNTL_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_LIMITS_H 1
#define HAVE_STDINT_H 1
#define HAVE_TERMIOS_H 1
#define HAVE_UNISTD_H 1
#define HAVE_LINUX_IF_TUN_H 1
#define HAVE_LINUX_TCP_H 1
#define HAVE_SYS_IOCTL_H 1
#define HAVE_SYS_POLL_H 1
#define HAVE_SYS_SOCKET_H 1
#define HAVE_SYS_SOUNDCARD_H 1
#define HAVE_SYS_TIME_H 1
#define HAVE_SYS_TYPES_H 1

#define HAVE_FSEEKO 1
#define HAVE_FTRUNCATE 1
#define HAVE_USLEEP 1
#define HAVE_NANOSLEEP 1
#define HAVE_SLEEP 1
#define HAVE_GETTIMEOFDAY 1

#define PCE_VERSION_MAJ 0
#define PCE_VERSION_MIN 2
#define PCE_VERSION_MIC 2
#define PCE_VERSION_STR "0.2.2"

#define PCE_HOST_LINUX 1
/* #undef PCE_HOST_WINDOWS */
/* #undef PCE_HOST_SUNOS */
/* #undef PCE_HOST_NETBSD */

/* #undef PCE_HOST_IA32 */
/* #undef PCE_HOST_PPC */
/* #undef PCE_HOST_SPARC */

#define PCE_DIR_ETC "NONE/etc"

#define PCE_BUILD_IBMPC 1

#define PCE_ENABLE_X11 1

#define PCE_ENABLE_SDL 1

/* #undef PCE_ENABLE_READLINE */

#define PCE_ENABLE_TUN 1

#define PCE_ENABLE_CHAR_POSIX 1
#define PCE_ENABLE_CHAR_PPP 1
#define PCE_ENABLE_CHAR_PTY 1
#define PCE_ENABLE_CHAR_SLIP 1
#define PCE_ENABLE_CHAR_TCP 1
#define PCE_ENABLE_CHAR_TIOS 1

#define PCE_ENABLE_SOUND_OSS 1

/* directory separator */
#define PCE_DIR_SEP '/'


#endif


src/arch/ibmpc/main.h
/*****************************************************************************
 * pce                                                                       *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/arch/ibmpc/main.h                                        *
 * Created:     2001-05-01 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2001-2011 Hampa Hug <hampa@hampa.ch>                     *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#ifndef PCE_IBMPC_MAIN_H
#define PCE_IBMPC_MAIN_H 1


#include <config.h>


#define PCE_IBMPC_CLK0 14318184
#define PCE_IBMPC_CLK1 (PCE_IBMPC_CLK0 / 3)
#define PCE_IBMPC_CLK2 (PCE_IBMPC_CLK0 / 12)

#define PCE_IBMPC_5150 1
#define PCE_IBMPC_5160 2
#define PCE_IBMPC_M24  4


extern const char *par_terminal;
extern const char *par_video;


void pc_log_deb (const char *msg, ...);


#endif


src/drivers/pfdc/pfdc-img.h
/*****************************************************************************
 * pce                                                                       *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/drivers/pfdc/pfdc-img.h                                  *
 * Created:     2012-02-14 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2012 Hampa Hug <hampa@hampa.ch>                          *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#ifndef PCE_DRV_PFDC_IMG_H
#define PCE_DRV_PFDC_IMG_H 1


#include <drivers/pfdc/pfdc.h>


#define PFDC_FORMAT_NONE    0
#define PFDC_FORMAT_ANADISK 1
#define PFDC_FORMAT_CP2     2
#define PFDC_FORMAT_DC42    3
#define PFDC_FORMAT_IMD     4
#define PFDC_FORMAT_PFDC    5
#define PFDC_FORMAT_PFDC0   6
#define PFDC_FORMAT_PFDC1   7
#define PFDC_FORMAT_PFDC2   8
#define PFDC_FORMAT_PFDC4   9
#define PFDC_FORMAT_RAW     10
#define PFDC_FORMAT_TC      11
#define PFDC_FORMAT_TD0     12
#define PFDC_FORMAT_XDF     13


unsigned pfdc_guess_type (const char *fname);

pfdc_img_t *pfdc_load_fp (FILE *fp, unsigned type);
pfdc_img_t *pfdc_load (const char *fname, unsigned type);

int pfdc_save_fp (FILE *fp, const pfdc_img_t *img, unsigned type);
int pfdc_save (const char *fname, const pfdc_img_t *img, unsigned type);

unsigned pfdc_probe_fp (FILE *fp);
unsigned pfdc_probe (const char *fname);


#endif


src/arch/macplus/main.c
/*****************************************************************************
 * pce                                                                       *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/arch/macplus/main.c                                      *
 * Created:     2007-04-15 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2007-2013 Hampa Hug <hampa@hampa.ch>                     *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#include "main.h"
#include "cmd_68k.h"
#include "macplus.h"
#include "msg.h"
#include "sony.h"

#include <stdarg.h>
#include <time.h>

#include <unistd.h>
#include <signal.h>

#include <lib/cmd.h>
#include <lib/console.h>
#include <lib/getopt.h>
#include <lib/log.h>
#include <lib/monitor.h>
#include <lib/path.h>
#include <lib/sysdep.h>

#include <libini/libini.h>

#ifdef PCE_ENABLE_SDL
#include <SDL.h>
#endif


const char *par_terminal = NULL;

unsigned   par_disk_delay_valid = 0;
unsigned   par_disk_delay[SONY_DRIVES];

macplus_t  *par_sim = NULL;

unsigned   par_sig_int = 0;

monitor_t  par_mon;

ini_sct_t  *par_cfg = NULL;

static ini_strings_t par_ini_str;


static pce_option_t opts[] = {
	{ '?', 0, "help", NULL, "Print usage information" },
	{ 'b', 1, "disk-delay-1", "delay", "Set the disk delay for drive 1 [30]" },
	{ 'B', 2, "disk-delay", "drive delay", "Set the disk delay [30]" },
	{ 'c', 1, "config", "string", "Set the config file name [none]" },
	{ 'd', 1, "path", "string", "Add a directory to the search path" },
	{ 'i', 1, "ini-prefix", "string", "Add an ini string before the config file" },
	{ 'I', 1, "ini-append", "string", "Add an ini string after the config file" },
	{ 'l', 1, "log", "string", "Set the log file name [none]" },
	{ 'p', 1, "cpu", "string", "Set the CPU model" },
	{ 'q', 0, "quiet", NULL, "Set the log level to error [no]" },
	{ 'r', 0, "run", NULL, "Start running immediately [no]" },
	{ 'R', 0, "no-monitor", NULL, "Never stop running [no]" },
	{ 's', 1, "speed", "int", "Set the CPU speed" },
	{ 't', 1, "terminal", "string", "Set the terminal device" },
	{ 'v', 0, "verbose", NULL, "Set the log level to debug [no]" },
	{ 'V', 0, "version", NULL, "Print version information" },
	{  -1, 0, NULL, NULL, NULL }
};


static
void print_help (void)
{
	pce_getopt_help (
		"pce-macplus: Macintosh Plus emulator",
		"usage: pce-macplus [options]",
		opts
	);

	fflush (stdout);
}

static
void print_version (void)
{
	fputs (
		"pce-macplus version " PCE_VERSION_STR
		"\n\n"
		"Copyright (C) 2007-2012 Hampa Hug <hampa@hampa.ch>\n",
		stdout
	);

	fflush (stdout);
}

static
void mac_log_banner (void)
{
	pce_log (MSG_INF,
		"pce-macplus version " PCE_VERSION_STR "\n"
		"Copyright (C) 2007-2012 Hampa Hug <hampa@hampa.ch>\n"
	);
}

void sig_int (int s)
{
	par_sig_int = 1;
}

void sig_segv (int s)
{
	pce_set_fd_interactive (0, 1);

	fprintf (stderr, "pce-macplus: segmentation fault\n");
	fflush (stderr);

	if ((par_sim != NULL) && (par_sim->cpu != NULL)) {
		mac_prt_state (par_sim, "cpu");
	}

	exit (1);
}

void sig_term (int s)
{
	if (par_sim != NULL) {
		mac_del (par_sim);
	}

	pce_set_fd_interactive (0, 1);

	fprintf (stderr, "pce-macplus: signal %d\n", s);
	fflush (stderr);

	exit (1);
}

static
void mac_atexit (void)
{
	pce_set_fd_interactive (0, 1);
}

static
int cmd_get_sym (macplus_t *sim, const char *sym, unsigned long *val)
{
	if (e68_get_reg (sim->cpu, sym, val) == 0) {
		return (0);
	}

	return (1);
}

static
int cmd_set_sym (macplus_t *sim, const char *sym, unsigned long val)
{
	if (e68_set_reg (sim->cpu, sym, val) == 0) {
		return (0);
	}

	return (1);
}

void mac_log_deb (const char *msg, ...)
{
	va_list       va;
	unsigned long pc;

	if (par_sim != NULL) {
		pc = e68_get_pc (par_sim->cpu);
	}
	else {
		pc = 0;
	}

	pce_log (MSG_DEB, "[%06lX] ", pc & 0xffffff);

	va_start (va, msg);
	pce_log_va (MSG_DEB, msg, va);
	va_end (va);
}

static
int pce_load_config (ini_sct_t *ini, const char *fname)
{
	if (fname == NULL) {
		return (0);
	}

	pce_log_tag (MSG_INF, "CONFIG:", "file=\"%s\"\n", fname);

	if (ini_read (par_cfg, fname)) {
		pce_log (MSG_ERR, "*** loading config file failed\n");
		return (1);
	}

	return (0);
}

int main (int argc, char *argv[])
{
	int       r;
	char      **optarg;
	int       run, nomon;
	unsigned  drive;
	char      *cfg;
	ini_sct_t *sct;

	cfg = NULL;
	run = 0;
	nomon = 0;

	pce_log_init();
	pce_log_add_fp (stderr, 0, MSG_INF);

	par_cfg = ini_sct_new (NULL);

	if (par_cfg == NULL) {
		return (1);
	}

	ini_str_init (&par_ini_str);

	while (1) {
		r = pce_getopt (argc, argv, &optarg, opts);

		if (r == GETOPT_DONE) {
			break;
		}

		if (r < 0) {
			return (1);
		}

		switch (r) {
		case '?':
			print_help();
			return (0);

		case 'V':
			print_version();
			return (0);

		case 'b':
			par_disk_delay_valid |= 1;
			par_disk_delay[0] = (unsigned) strtoul (optarg[0], NULL, 0);
			break;

		case 'B':
			drive = strtoul (optarg[0], NULL, 0);

			if ((drive < 1) || (drive >= SONY_DRIVES)) {
				fprintf (stderr, "%s: bad drive number (%u)\n",
					argv[0], drive
				);
				return (1);
			}

			drive -= 1;

			par_disk_delay_valid |= 1U << drive;
			par_disk_delay[drive] = (unsigned) strtoul (optarg[1], NULL, 0);
			break;

		case 'c':
			cfg = optarg[0];
			break;

		case 'd':
			pce_path_set (optarg[0]);
			break;

		case 'i':
			if (ini_read_str (par_cfg, optarg[0])) {
				fprintf (stderr,
					"%s: error parsing ini string (%s)\n",
					argv[0], optarg[0]
				);
				return (1);
			}
			break;

		case 'I':
			ini_str_add (&par_ini_str, optarg[0], "\n", NULL);
			break;

		case 'l':
			pce_log_add_fname (optarg[0], MSG_DEB);
			break;

		case 'p':
			ini_str_add (&par_ini_str, "cpu.model = \"",
				optarg[0], "\"\n"
			);
			break;

		case 'q':
			pce_log_set_level (stderr, MSG_ERR);
			break;

		case 'r':
			run = 1;
			break;

		case 'R':
			nomon = 1;
			break;

		case 's':
			ini_str_add (&par_ini_str, "cpu.speed = ",
				optarg[0], "\n"
			);
			break;

		case 't':
			par_terminal = optarg[0];
			break;

		case 'v':
			pce_log_set_level (stderr, MSG_DEB);
			break;

		case 0:
			fprintf (stderr, "%s: unknown option (%s)\n",
				argv[0], optarg[0]
			);
			return (1);

		default:
			return (1);
		}
	}

	mac_log_banner();

	if (pce_load_config (par_cfg, cfg)) {
		return (1);
	}

	sct = ini_next_sct (par_cfg, NULL, "macplus");

	if (sct == NULL) {
		sct = par_cfg;
	}

	if (ini_str_eval (&par_ini_str, sct, 1)) {
		return (1);
	}

	atexit (mac_atexit);

#ifdef PCE_ENABLE_SDL
	SDL_Init (0);
#endif

	pce_path_ini (sct);

	signal (SIGINT, &sig_int);
	signal (SIGSEGV, &sig_segv);
	signal (SIGTERM, &sig_term);

	pce_console_init (stdin, stdout);

	par_sim = mac_new (sct);

	mon_init (&par_mon);
	mon_set_cmd_fct (&par_mon, mac_cmd, par_sim);
	mon_set_msg_fct (&par_mon, mac_set_msg, par_sim);
	mon_set_get_mem_fct (&par_mon, par_sim->mem, mem_get_uint8);
	mon_set_set_mem_fct (&par_mon, par_sim->mem, mem_set_uint8);
	mon_set_memory_mode (&par_mon, 0);

	cmd_init (par_sim, cmd_get_sym, cmd_set_sym);
	mac_cmd_init (par_sim, &par_mon);

	mac_reset (par_sim);

	if (nomon) {
		while (par_sim->brk != PCE_BRK_ABORT) {
			mac_run (par_sim);
		}
	}
	else if (run) {
		mac_run (par_sim);
		if (par_sim->brk != PCE_BRK_ABORT) {
			pce_puts ("\n");
		}
	}
	else {
		pce_puts ("type 'h' for help\n");
	}

	if (par_sim->brk != PCE_BRK_ABORT) {
		mon_run (&par_mon);
	}

	mac_del (par_sim);

#ifdef PCE_ENABLE_SDL
	SDL_Quit();
#endif

	mon_free (&par_mon);
	pce_console_done();
	pce_log_done();

	return (0);
}
